### PR TITLE
issue : #88 and #90

### PR DESCRIPTION
**Pull request by salarSb**

### CODE REVIEW

File: database/migrations/2022_05_16_094029_change_tags_column_type_in_ads.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::table('ads', function (Blueprint $table) {
+            //
```
Comment (amirpourastarabadi) [inline database/migrations/2022_05_16_094029_change_tags_column_type_in_ads.php]: write down method please.

File: app/Repositories/Repo.php [hunk 2]
```diff
@@ -19,8 +19,9 @@ class Repo
      * @param array $multiValueFilter
      * @return Builder
      */
-    public static function getRecords(string $table, array $columns=[] , array $filters=[] , array $multiValueFilter=[]):Builder{
-        if(empty($columns) || in_array('*' , $columns))
+    public static function getRecords(string $table, array $columns = [], array $filters = [], array $multiValueFilter = []): Builder
```
Comment (amirpourastarabadi) [inline app/Repositories/Repo.php:21]: as $filters and $multiValueFilter variables have no role in getRecords body, you can remove them

File: app/Repositories/Repo.php [hunk 3]
```diff
@@ -31,40 +32,43 @@ public static function getRecords(string $table, array $columns=[] , array $filt
...
+    public static function updateRecord(string $table, int $id, array $data): bool
     {
-        return DB::table($table)->update($data);
+        return DB::table($table)->where('id', $id)->update($data);
     }
 
-    public static function valid(Builder $query):Builder{
-        return $query->where(function (Builder $q){
-            $q ->where('since', '<=', Carbon::createFromFormat('Y-m-d H:i:s', Carbon::now('Asia/Tehran')))
+    public static function valid(Builder $query): Builder
```
Comment (amirpourastarabadi) [inline app/Repositories/Repo.php]: No use case in project! please remove it.

File: app/Repositories/Repo.php [hunk 4]
```diff
@@ -31,40 +32,43 @@ public static function getRecords(string $table, array $columns=[] , array $filt
...
-        })->where(function (Builder $q){
-            $q ->where('till', '>=', Carbon::createFromFormat('Y-m-d H:i:s', Carbon::now('Asia/Tehran')))
+        })->where(function (Builder $q) {
+            $q->where('till', '>=', Carbon::createFromFormat('Y-m-d H:i:s', Carbon::now('Asia/Tehran')))
                 ->orWhereNull('till');
         });
     }
 
-    public static function enable(Builder $query) :Builder
+    public static function enable(Builder $query): Builder
```
Comment (amirpourastarabadi) [inline app/Repositories/Repo.php]: No use case in project! please remove it.

File: app/Jobs/FetchAd.php [hunk 5]
```diff
@@ -37,28 +45,27 @@ public function __construct(string $sourceName)
      * @param AdItemInserter $adItemInserter
      * @param AdPicTransferrer $adPicTransferrer
      */
-    public function handle(AdFetcher $adFetcher , AdItemInserter $adItemInserter , AdPicTransferrer $adPicTransferrer)
+    public function handle(AdFetcher $adFetcher, AdItemInserter $adItemInserter, AdPicTransferrer $adPicTransferrer)
     {
         $this->adFetcher = $adFetcher;
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php:50]: set adFetcher in counstructor via Ioc Container(like handle method)

Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php:50]: set adFetcher in counstructor via Ioc Container(like handle method)

File: app/Jobs/FetchAd.php [hunk 6]
```diff
@@ -37,28 +45,27 @@ public function __construct(string $sourceName)
...
-        $this->adItemInserter=$adItemInserter;
-        $this->adPicTransferrer=$adPicTransferrer;
+        $this->adItemInserter = $adItemInserter;
+        $this->adPicTransferrer = $adPicTransferrer;
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php]: set adFetcher, adItemInserter, adPicTransferrer in counstructor via Ioc Container(like handle method)

File: app/Jobs/FetchAd.php [hunk 7]
```diff
@@ -37,28 +45,27 @@ public function __construct(string $sourceName)
...
 
         /** @var stdClass $source */
-        $source = Repo::getRecords('sources' , ['*'], ['name'=>$this->sourceName])->first();
-        if(isset($source)){
-            [$donePages, $failedPages] = $this->fetch($source);
+        if (isset($this->source)) {
+            [$donePages, $failedPages] = $this->fetch($this->source, $this->since);
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php]: remove $source and since from fetch method signature.

File: app/Jobs/FetchAd.php [hunk 8]
```diff
@@ -67,50 +74,45 @@ private function fetch(stdClass $source): array
...
-            [$fetchDone , $items , $currentPage , $nextPageUrl , $lastPage, $resultText] = $this->adFetcher->fetchAd($fetchUrl);
-            if(!$fetchDone){
+            [$fetchDone, $items, $currentPage, $nextPageUrl, $lastPage, $resultText] = $this->adFetcher->fetchAd($fetchUrl);
+            if (!$fetchDone) {
                 $failedPages++;
                 continue;
             }
 
             if (empty($items)) {
                 continue;
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php:113]: It may cause infinite loop, put some think behind it please.

File: app/Jobs/FetchAd.php [hunk 9]
```diff
@@ -67,50 +74,45 @@ private function fetch(stdClass $source): array
...
             }
-
             $this->storeItems($source, $items, $counter);
 
-            $this->insertOrUpdateFetch($source, $currentPage, $lastPage, $nextPageUrl);
+            if ($currentPage == 1) {
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php]: use Identical operator here(===)

File: database/migrations/2022_05_16_094029_change_tags_column_type_in_ads.php (hunk 1 above)
Comment (salarSb) [reply database/migrations/2022_05_16_094029_change_tags_column_type_in_ads.php]: done

File: app/Repositories/Repo.php (hunk 2 above)
Comment (salarSb) [reply app/Repositories/Repo.php:21]: they are passed to filter and filterMultipleValue static functions to use where and whereIn in queries

File: app/Repositories/Repo.php (hunk 3 above)
Comment (salarSb) [reply app/Repositories/Repo.php]: done

File: app/Repositories/Repo.php (hunk 4 above)
Comment (salarSb) [reply app/Repositories/Repo.php]: done

File: app/Jobs/FetchAd.php (hunk 7 above)
Comment (salarSb) [reply app/Jobs/FetchAd.php]: done

File: app/Jobs/FetchAd.php (hunk 9 above)
Comment (salarSb) [reply app/Jobs/FetchAd.php]: done

File: app/Jobs/FetchAd.php (hunk 8 above)
Comment (salarSb) [reply app/Jobs/FetchAd.php:113]: done

File: app/Jobs/FetchAd.php (hunk 6 above)
Comment (salarSb) [reply app/Jobs/FetchAd.php]: done

File: app/Jobs/FetchAd.php [hunk 10]
```diff
@@ -9,108 +9,114 @@
...
     {
-        $this->adFetcher = $adFetcher;
-        $this->adItemInserter=$adItemInserter;
-        $this->adPicTransferrer=$adPicTransferrer;
-
         /** @var stdClass $source */
-        $source = Repo::getRecords('sources' , ['*'], ['name'=>$this->sourceName])->first();
-        if(isset($source)){
-            [$donePages, $failedPages] = $this->fetch($source);
+        if (isset($this->source)) {
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php:57]: use early return in if else structure if it is possible

File: app/Jobs/FetchAd.php [hunk 11]
```diff
@@ -9,108 +9,114 @@
...
-        $failedPages = 0;
         $donePages = 0;
 
         do {
             $counter = 0;
-            [$fetchDone , $items , $currentPage , $nextPageUrl , $lastPage, $resultText] = $this->adFetcher->fetchAd($fetchUrl);
-            if(!$fetchDone){
-                $failedPages++;
-                continue;
+            [$fetchDone, $items, $currentPage, $nextPageUrl, $lastPage, $resultText] = $this->adFetcher->fetchAd($fetchUrl);
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php:82]: refactor this array to a class structure and set this variables in a class properties base your ideas

File: app/Jobs/FetchAd.php [hunk 12]
```diff
@@ -9,108 +9,114 @@
...
+            if (!$fetchDone) {
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php:97]: iterate on all urls if previous response class and send request to get ads.
store failed requests to some data structure and after finishing first iteration try for 3 times on failed urls with 5 min sleep.
finally log failed urls and exit

File: app/Jobs/FetchAd.php [hunk 13]
```diff
@@ -9,108 +9,114 @@
...
+                Log::error('response status code is not 200 | request url: ' . $fetchUrl);
+                throw new Exception('response status code is not 200 | request url: ' . $fetchUrl);
             }
 
             if (empty($items)) {
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php]: like last comment log empty items and continue to next url

File: app/Jobs/FetchAd.php [hunk 14]
```diff
@@ -9,108 +9,114 @@
...
-                continue;
+                Log::error('response data is null | request url: ' . $fetchUrl);
+                throw new Exception('response data is null | request url: ' . $fetchUrl);
+            }
+            $this->storeItems($this->source, $items, $counter);
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php]: if $counter is redundant remove it

File: app/Jobs/FetchAd.php [hunk 15]
```diff
@@ -9,108 +9,114 @@
...
-     * @param $lastPage
-     * @param $nextPageUrl
-     * @return bool
-     */
-    private function insertOrUpdateFetch(stdClass $source, $currentPage, $lastPage, $nextPageUrl): bool
-    {
-        if ($currentPage == 1) {
-            return $this->insertFetch($source->id, $currentPage, $lastPage, $nextPageUrl);
-        }
+            if ($currentPage === $this->lastFetch->last_page) {
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php:132]: check in this if that failed urls are empty and if true set completed_at for fetch

File: app/Jobs/FetchAd.php [hunk 16]
```diff
@@ -9,108 +9,114 @@
...
             }
 
             if (empty($items)) {
-                continue;
+                Log::error('response data is null | request url: ' . $fetchUrl);
+                throw new Exception('response data is null | request url: ' . $fetchUrl);
+            }
+            $this->storeItems($this->source, $items, $counter);
+
+            if ($currentPage === 1) {
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php:122]: refactor: 
1- remove do while loop 
2- create fetch with model if first request was success
3- update fetch in for loop if request was success

File: app/Repositories/Repo.php [hunk 17]
```diff
@@ -4,7 +4,6 @@
 namespace App\Repositories;
```
Comment (amirpourastarabadi) [inline app/Repositories/Repo.php:4]: remove this Repo and use model

File: app/Jobs/FetchAd.php [hunk 18]
```diff
@@ -9,108 +9,114 @@
...
      * @param int $sourceID
      * @param int $currentPage
-     * @param int $lastPage
-     * @param string $nextPageUrl
+     * @param int|null $lastPage
+     * @param string|null $nextPageUrl
      * @return bool
      */
-    private function insertFetch(int $sourceID, int $currentPage, int $lastPage=null, string $nextPageUrl=null):bool
+    private function insertFetch(int $sourceID, int $currentPage, int $lastPage = null, string $nextPageUrl = null): bool
```
Comment (amirpourastarabadi) [inline app/Jobs/FetchAd.php:150]: remove all this method that have relation to database activity

File: app/Http/Controllers/HomeController.php [hunk 19]
```diff
@@ -26,61 +27,63 @@ public function debug(Request $request){
...
-        $customer = Repo::getRecords('users', ['*'], ['UUID'=>$customerUUID])->first();
-        if(is_null($customer)){
+    public function getAd(Request $request, AdCollector $adResponseGenerator)
+    {
+        $numberOfAds = $request->input('numberOfAds', 6);
+        $customerUUID = $request->input('UUID');
+        $sourceNames = $request->input('source', []);
+        $urls = $request->input('urls', []);
+
+        $customer = Repo::getRecords('users', ['*'], ['UUID' => $customerUUID])->first();
```
Comment (amirpourastarabadi) [inline app/Http/Controllers/HomeController.php:37]: Refactor base Repo removing

File: app/Http/Controllers/HomeController.php [hunk 20]
```diff
@@ -26,61 +27,63 @@ public function debug(Request $request){
...
+        if (is_null($customer)) {
             return response()->json($this->setErrorResponse(myResponse::USER_NOT_FOUND, 'User not found'));
         }
 
-        if(is_string($sourceNames)){
+        if (is_string($sourceNames)) {
```
Comment (amirpourastarabadi) [inline app/Http/Controllers/HomeController.php:42]: use middleware

File: app/Http/Controllers/HomeController.php [hunk 21]
```diff
@@ -26,61 +27,63 @@ public function debug(Request $request){
...
             $sourceNames = convertTagStringToArray($sourceNames);
         }
 
-        $sources = SourceRepo::getValidSource($customer->id,$sourceNames,[])->get();
-        if($sources->isEmpty()){
+        $sources = SourceRepo::getValidSource($customer->id, $sourceNames, [])->get();
```
Comment (amirpourastarabadi) [inline app/Http/Controllers/HomeController.php:46]: remove SourceRepo and use model instead

File: app/Http/Controllers/HomeController.php [hunk 22]
```diff
@@ -26,61 +27,63 @@ public function debug(Request $request){
...
-        $ad = Repo::getRecords('ads', ['*'], [$this->makeAdForeignId($source->id , $itemID , $itemType)])->first();
-
-        if(is_null($ad)){
+    public function fetchAd(Request $request)
+    {
+        //ToDo : security alert : anyone can update Chibekhoonam ads
+        $itemID = $request->input('item_id');
+        $itemType = $request->input('item_type');
+        $sourceName = $request->input('source');
+        $source = Repo::getRecords('sources', ['*'], ['name' => $sourceName])->first();
```
Comment (amirpourastarabadi) [inline app/Http/Controllers/HomeController.php:61]: refactor base Repo removing

File: app/Http/Controllers/HomeController.php [hunk 23]
```diff
@@ -26,61 +27,63 @@ public function debug(Request $request){
...
-    public function fetchAd(Request $request){
-        //ToDo : security alert : any one can update Chibekhoonam ads
-        $itemID         = $request->input('item_id');
-        $itemType       = $request->input('item_type');
-        $sourceName     = $request->input('source');
-        $source = Repo::getRecords('sources', ['*'], ['name'=>$sourceName])->first();
-        $ad = Repo::getRecords('ads', ['*'], [$this->makeAdForeignId($source->id , $itemID , $itemType)])->first();
-
-        if(is_null($ad)){
+    public function fetchAd(Request $request)
```
Comment (amirpourastarabadi) [inline app/Http/Controllers/HomeController.php:55]: rename method name

File: app/Http/Controllers/HomeController.php [hunk 24]
```diff
@@ -26,61 +27,63 @@ public function debug(Request $request){
...
+        $ad = Repo::getRecords('ads', ['*'], ['foreign_id' => $this->makeAdForeignId($source->id, $itemID, $itemType),])->first();
+        if (is_null($ad)) {
             return response()->json($this->setErrorResponse(myResponse::AD_NOT_FOUND, 'Ad not found'));
         }
 
-        $update = DB::table('ads')->update([
-           'name'   => $request->input('name' , optional($ad)->name),
-           'image'  => $request->input('link' , optional($ad)->link),
-           'link'   => $request->input('image' , optional($ad)->image),
+        $update = DB::table('ads')->where('id', $ad->id)->update([
```
Comment (amirpourastarabadi) [inline app/Http/Controllers/HomeController.php:67]: update base model

File: app/Http/Controllers/HomeController.php [hunk 25]
```diff
@@ -26,61 +27,63 @@ public function debug(Request $request){
...
         }
 
         return response()->json($this->setErrorResponse(myResponse::AD_UPDATE_DATABASE_ERROR, 'Database error on updating ad'));
     }
 
-    public function adClick(Request $request , string $UUID){
-        $ad = Repo::getRecords('ads' , ['*'] , ['UUID'=>$UUID])->first();
+    public function adClick(Request $request, string $UUID)
+    {
+        $ad = Repo::getRecords('ads', ['*'], ['UUID' => $UUID])->first();
```
Comment (amirpourastarabadi) [inline app/Http/Controllers/HomeController.php:84]: refactor base Repo removing

File: app/Http/Controllers/HomeController.php [hunk 26]
```diff
@@ -26,61 +27,63 @@ public function debug(Request $request){
...
         Ad::setReferer($ad, $request->headers->get('referer'));
```
Comment (amirpourastarabadi) [inline app/Http/Controllers/HomeController.php:85]: find the role of this line please

File: app/Console/Kernel.php [hunk 27]
```diff
@@ -19,12 +21,16 @@ class Kernel extends ConsoleKernel
...
+     * @param \Illuminate\Console\Scheduling\Schedule $schedule
      * @return void
      */
     protected function schedule(Schedule $schedule)
     {
-        // $schedule->command('inspire')->hourly();
+        $since = Carbon::now()->subMonths(3)->toDateString();
+//        cron('0 0 * */3 *')
+        //TODO: enter the exact sourceName of chibekhunam instead of test
+        $schedule->command(Fetching::class, ['test', '--since=' . $since])
```
Comment (amirpourastarabadi) [inline app/Console/Kernel.php]: loop throw all distinct sources

File: app/Console/Kernel.php [hunk 28]
```diff
@@ -19,12 +21,16 @@ class Kernel extends ConsoleKernel
...
+            ->everyMinute()->timezone('Asia/Tehran');
```
Comment (amirpourastarabadi) [inline app/Console/Kernel.php]: every 3 months

File: config/filesystems.php [hunk 29]
```diff
@@ -46,47 +46,20 @@
...
-            'privateKey' => env('SFTP_PRIVATE_KEY_PATH', ''),
-            'root'       => env('SFTP_ROOT',''),
-            'timeout'    => env('SFTP_TIMEOUT', '10'),
-            'prefix'     => null,
-            'dHost'      => "cdn.alaatv.com",
-            'dProtocol'  => "https://",
-        ],
-
-
-        's3' => [
```
Comment (amirpourastarabadi) [inline config/filesystems.php:83]: this name is general select a special name for ads

File: app/Classes/AdPicTransferrer.php [hunk 30]
```diff
@@ -18,22 +18,22 @@ class AdPicTransferrer
...
     {
-        try{
-            if(!isset($picUrl))
+        try {
+            if (!isset($picUrl))
                 return [false, null];
 
-            $pathToSave = Storage::disk('adImage')->getAdapter()->getPathPrefix() . basename($picUrl);
-            $filePath = fopen($pathToSave, 'w');
+            $pathToSave = Storage::disk('adImage')->putFileAs('', $picUrl, basename($picUrl));
```
Comment (amirpourastarabadi) [inline app/Classes/AdPicTransferrer.php]: refactor this base `Boundaries` chapter of `clean code book`

File: config/filesystems.php [hunk 31]
```diff
@@ -29,7 +29,7 @@
     */
 
 
-    'cloud' => env('FILESYSTEM_CLOUD', 's3'),
+    'cloud' => env('FILESYSTEM_CLOUD', 'adImageMinio'),
```
Comment (amirpourastarabadi) [inline config/filesystems.php]: set default to s3

File: app/Classes/AdImageUtil.php [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+    public function uploadImageToLocal(string $picUrl): static
+    {
+        $this->pathToSave = Storage::putFileAs('', $picUrl, $this->fileName);
+        return $this;
+    }
+
+    public function uploadImageToMinio($filePath): bool
+    {
+        return Storage::cloud()->put($this->fileName, File::get(storage_path($filePath)));
+    }
```
Comment (amirpourastarabadi) [inline app/Classes/AdImageUtil.php]: get file as parameter not from storage

File: app/Classes/AdItemInserter.php [hunk 33]
```diff
@@ -81,29 +85,34 @@ private function updateAdRecord($item)
...
      * @param AdPicTransferrer $adPicTransferrer
      * @return array
      */
-    private function putAdPicToCDN($item, AdPicTransferrer $adPicTransferrer): array
+    private function putAdPicToMinio($item, AdPicTransferrer $adPicTransferrer): array
     {
         $isPicTransferred = false;
-        $picUrl=null;
+        $picUrl = null;
         [$storeResult, $picPath] = $adPicTransferrer->storeAdPic(optional($item)->image);
```
Comment (amirpourastarabadi) [inline app/Classes/AdItemInserter.php]: store file directly to minio.(we do not need store it in local storage)

**Review by amirpourastarabadi — APPROVED** [review]

**Pull request closed without merging**